[BUG] linked strategy: `npm install` leaves dangling symlinks after store hash recalculation

## Is there an existing issue for this?

- [x] I have searched the existing issues

## This issue exists in the latest npm version

- [x] I am using the latest npm

## Current Behavior

When using `install-strategy=linked`, running `npm install --save-dev <new-package>` after an initial install leaves dangling symlinks in `node_modules/` for existing packages whose store hashes changed.

The store hash (the suffix in `.store/pkg@version-HASH/`) is recalculated when the dependency graph changes. npm creates new `.store/` directories with updated hashes and removes the old ones, but does not update the symlinks that still reference the old hash paths. This breaks `node_modules/.bin/` entries and any `require()` calls for the affected packages.

### Discovery

Found while testing the last release (v10.9.6) in [Gutenberg](https://github.com/WordPress/gutenberg/pull/75814). After `npm ci` followed by `npm install --save-dev storybook@^10.1.11 @storybook/test-runner@0.24.2 playwright`, the `wait-on` binary became unresolvable due to a dangling symlink — see [CI failure](https://github.com/WordPress/gutenberg/actions/runs/23003481169/job/66793810678).

## Expected Behavior

After `npm install --save-dev`, all symlinks in `node_modules/` and `node_modules/.bin/` should point to valid targets in `.store/`.

## Root Cause

The store hash is calculated in [`getKey`](https://github.com/npm/cli/blob/3b70a9d1ade5aa5e7edebd0dbc84f0c5f6f202a9/workspaces/arborist/lib/arborist/isolated-reifier.js#L9) by hashing the **entire dependency subtree** — package names, versions, resolved URLs, and their full dependency paths. Adding any package changes the graph, which changes hashes for existing packages.

During reification:

1. [`#buildLinkedActualForDiff`](https://github.com/npm/cli/blob/c7702d044907e56e06c769fa7a42ed1ed308f29a/workspaces/arborist/lib/arborist/reify.js#L807) builds a synthetic actual tree from the ideal tree's children rather than reading the actual filesystem state. This means the synthetic "actual" already has the new resolved paths (containing new hashes), so the diff engine sees `ideal.resolved === actual.resolved` and returns no action needed.

2. [`#cleanOrphanedStoreEntries`](https://github.com/npm/cli/blob/c7702d044907e56e06c769fa7a42ed1ed308f29a/workspaces/arborist/lib/arborist/reify.js#L1317) then removes old `.store/` directories that are no longer in the ideal tree.

3. Result: symlinks still point to the old (now deleted) `.store/` paths → dangling.

The key issue is that the diff comparison ([`getAction`](https://github.com/npm/cli/blob/26fa40eeafdbbb616d48fe254c92544cb13fba60/workspaces/arborist/lib/diff.js#L112)) cannot detect that symlinks need updating because `#buildLinkedActualForDiff` doesn't reflect actual on-disk symlink targets.

## Steps To Reproduce

```bash
mkdir test-dangling && cd test-dangling

cat > package.json << 'EOF'
{
  "name": "test-dangling",
  "version": "1.0.0",
  "devDependencies": {
    "wait-on": "^8.0.1"
  }
}
EOF

echo "install-strategy=linked" > .npmrc

# 1. Install
npm install

# 2. Verify symlink works
readlink node_modules/wait-on
# → .store/wait-on@8.0.5-RRJvnGlW6s4zr0GHYWoIKg/node_modules/wait-on  ✅
node -e "require('wait-on'); console.log('OK')"  # ✅

# 3. Install an additional package (changes the dependency graph)
npm install --save-dev express@4

# 4. Symlink is now dangling
readlink node_modules/wait-on
# → .store/wait-on@8.0.5-RRJvnGlW6s4zr0GHYWoIKg/node_modules/wait-on  (old hash, unchanged)

ls node_modules/.store/ | grep wait-on
# wait-on@8.0.5-PnUP7Imr-963_DBuS2pz3A  (NEW hash — old dir deleted)

node -e "require('wait-on')"
# Error: Cannot find module 'wait-on'  ❌
```

## Environment

Reproduced on both v10 and latest:

- npm: 10.9.6 (`release/v10`) and 11.11.1 (`latest`)
- Node.js: v22.20.0
- OS Name: macOS (Darwin 25.3.0), Ubuntu 24.04
- npm config:

```ini
install-strategy=linked
```

--- a/workspaces/arborist/lib/arborist/reify.js
+++ b/workspaces/arborist/lib/arborist/reify.js
@@ -10,7 +10,7 @@ const { callLimit: promiseCallLimit } = require('promise-call-limit')
 const { depth: dfwalk } = require('treeverse')
 const { dirname, resolve, relative, join, sep } = require('node:path')
 const { log, time } = require('proc-log')
-const { existsSync } = require('node:fs')
+const { existsSync, readlinkSync } = require('node:fs')
 const { lstat, mkdir, readdir, rm, symlink } = require('node:fs/promises')
 const { moveFile } = require('@npmcli/fs')
 const { subset, intersects } = require('semver')
@@ -818,6 +818,17 @@ module.exports = cls => class Reifier extends cls {
       let entry
       if (child.isLink) {
         entry = new IsolatedLink(child)
+        // Read the on-disk symlink target so the diff detects store hash changes.
+        if (child.resolved?.startsWith('file:.store/')) {
+          try {
+            const onDiskTarget = readlinkSync(child.path)
+            if (onDiskTarget.startsWith('.store/')) {
+              entry.resolved = `file:${onDiskTarget}`
+            }
+          } catch {
+            // If we can't read the symlink, fall through to the ideal value
+          }
+        }
       } else {
         entry = new IsolatedNode(child)
       }

--- a/workspaces/arborist/lib/arborist/reify.js
+++ b/workspaces/arborist/lib/arborist/reify.js
@@ -10,7 +10,7 @@ const { callLimit: promiseCallLimit } = require('promise-call-limit')
 const { depth: dfwalk } = require('treeverse')
 const { dirname, resolve, relative, join, sep } = require('node:path')
 const { log, time } = require('proc-log')
-const { existsSync, readlinkSync } = require('node:fs')
+const { existsSync } = require('node:fs')
 const { lstat, mkdir, readdir, rm, symlink } = require('node:fs/promises')
 const { moveFile } = require('@npmcli/fs')
 const { subset, intersects } = require('semver')
@@ -815,26 +815,13 @@ module.exports = cls => class Reifier extends cls {
       if (combined.has(child.path) || !existsSync(child.path)) {
         continue
       }
+      // Skip store links whose ideal realpath doesn't exist on disk yet — the store hash changed and the symlink needs recreating via ADD.
+      if (child.isLink && child.resolved?.startsWith('file:.store/') && !existsSync(child.realpath)) {
+        continue
+      }
       let entry
       if (child.isLink) {
         entry = new IsolatedLink(child)
-        // Read the on-disk symlink target so the diff detects store hash changes. Only for root-level store links (path does not contain .store); store-internal links are recreated with their parent directory.
-        if (child.resolved?.startsWith('file:.store/') && !child.path.includes(`${sep}.store${sep}`)) {
-          try {
-            const onDiskTarget = readlinkSync(child.path)
-            // On POSIX, readlink returns ".store/pkg@ver-HASH/node_modules/pkg". On Windows, junctions return absolute paths like "D:\...\node_modules\.store\pkg@ver-HASH\...".
-            const storeMarker = `${sep}.store${sep}`
-            const storeIdx = onDiskTarget.indexOf(storeMarker)
-            if (onDiskTarget.startsWith(`.store${sep}`) || onDiskTarget.startsWith('.store/')) {
-              entry.resolved = `file:${onDiskTarget.split(sep).join('/')}`
-            } else /* istanbul ignore next -- Windows junctions return absolute paths */ if (storeIdx !== -1) {
-              const rel = onDiskTarget.substring(storeIdx + 1).split(sep).join('/')
-              entry.resolved = `file:${rel}`
-            }
-          } catch {
-            // If we can't read the symlink, fall through to the ideal value
-          }
-        }
       } else {
         entry = new IsolatedNode(child)
       }

--- a/workspaces/arborist/lib/arborist/reify.js
+++ b/workspaces/arborist/lib/arborist/reify.js
@@ -818,17 +818,16 @@ module.exports = cls => class Reifier extends cls {
       let entry
       if (child.isLink) {
         entry = new IsolatedLink(child)
-        // Read the on-disk symlink target so the diff detects store hash changes.
-        if (child.resolved?.startsWith('file:.store/')) {
+        // Read the on-disk symlink target so the diff detects store hash changes. Only for root-level store links (path does not contain .store); store-internal links are recreated with their parent directory.
+        if (child.resolved?.startsWith('file:.store/') && !child.path.includes(`${sep}.store${sep}`)) {
           try {
             const onDiskTarget = readlinkSync(child.path)
-            // On POSIX, readlink returns a relative path like ".store/pkg@ver-HASH/node_modules/pkg". On Windows, junctions store absolute paths like "D:\...\node_modules\.store\pkg@ver-HASH\...". Extract the ".store/..." portion in both cases.
+            // On POSIX, readlink returns ".store/pkg@ver-HASH/node_modules/pkg". On Windows, junctions return absolute paths like "D:\...\node_modules\.store\pkg@ver-HASH\...".
             const storeMarker = `${sep}.store${sep}`
             const storeIdx = onDiskTarget.indexOf(storeMarker)
-            if (onDiskTarget.startsWith('.store/') || onDiskTarget.startsWith(`.store${sep}`)) {
+            if (onDiskTarget.startsWith(`.store${sep}`) || onDiskTarget.startsWith('.store/')) {
               entry.resolved = `file:${onDiskTarget.split(sep).join('/')}`
-            /* istanbul ignore next -- Windows junctions return absolute paths */
-            } else if (storeIdx !== -1) {
+            } else /* istanbul ignore next -- Windows junctions return absolute paths */ if (storeIdx !== -1) {
               const rel = onDiskTarget.substring(storeIdx + 1).split(sep).join('/')
               entry.resolved = `file:${rel}`
             }

--- a/workspaces/arborist/lib/arborist/reify.js
+++ b/workspaces/arborist/lib/arborist/reify.js
@@ -822,8 +822,15 @@ module.exports = cls => class Reifier extends cls {
         if (child.resolved?.startsWith('file:.store/')) {
           try {
             const onDiskTarget = readlinkSync(child.path)
-            if (onDiskTarget.startsWith('.store/')) {
-              entry.resolved = `file:${onDiskTarget}`
+            // On POSIX, readlink returns a relative path like ".store/pkg@ver-HASH/node_modules/pkg". On Windows, junctions store absolute paths like "D:\...\node_modules\.store\pkg@ver-HASH\...". Extract the ".store/..." portion in both cases.
+            const storeMarker = `${sep}.store${sep}`
+            const storeIdx = onDiskTarget.indexOf(storeMarker)
+            if (onDiskTarget.startsWith('.store/') || onDiskTarget.startsWith(`.store${sep}`)) {
+              entry.resolved = `file:${onDiskTarget.split(sep).join('/')}`
+            /* istanbul ignore next -- Windows junctions return absolute paths */
+            } else if (storeIdx !== -1) {
+              const rel = onDiskTarget.substring(storeIdx + 1).split(sep).join('/')
+              entry.resolved = `file:${rel}`
             }
           } catch {
             // If we can't read the symlink, fall through to the ideal value

--- a/workspaces/arborist/test/isolated-mode.js
+++ b/workspaces/arborist/test/isolated-mode.js
@@ -1979,6 +1979,53 @@ tap.test('orphaned store entries are cleaned up on dependency removal', async t 
     'all store entries are removed when dependencies are removed')
 })
 
+tap.test('store symlinks are updated when hash changes after adding a dep', async t => {
+  const graph = {
+    registry: [
+      { name: 'which', version: '1.0.0', dependencies: { isexe: '^1.0.0' } },
+      { name: 'isexe', version: '1.0.0' },
+      { name: 'abbrev', version: '2.0.0' },
+    ],
+    root: {
+      name: 'myproject',
+      version: '1.0.0',
+      dependencies: { which: '1.0.0' },
+    },
+  }
+  const { dir, registry } = await getRepo(graph)
+  const cache = fs.mkdtempSync(`${getTempDir()}/test-`)
+
+  // First install — only which
+  const arb1 = new Arborist({ path: dir, registry, packumentCache: new Map(), cache })
+  await arb1.reify({ installStrategy: 'linked' })
+
+  const whichLink = path.join(dir, 'node_modules', 'which')
+  t.ok(fs.lstatSync(whichLink).isSymbolicLink(), 'which is a symlink after first install')
+  const hashBefore = fs.readlinkSync(whichLink)
+
+  // Add abbrev — changes the dep graph, causing store hash recalculation
+  const pkgPath = path.join(dir, 'package.json')
+  const pkg = JSON.parse(fs.readFileSync(pkgPath, 'utf8'))
+  pkg.dependencies.abbrev = '2.0.0'
+  fs.writeFileSync(pkgPath, JSON.stringify(pkg))
+
+  const arb2 = new Arborist({ path: dir, registry, packumentCache: new Map(), cache })
+  await arb2.reify({ installStrategy: 'linked' })
+
+  // The symlink target should still be valid (not dangling)
+  t.ok(fs.existsSync(whichLink), 'which symlink target exists after adding abbrev')
+  t.ok(setupRequire(dir)('which'), 'which is requireable after adding abbrev')
+
+  // Verify the symlink was updated if the hash changed
+  const hashAfter = fs.readlinkSync(whichLink)
+  if (hashBefore !== hashAfter) {
+    const storeDir = path.join(dir, 'node_modules', '.store')
+    const storeEntries = fs.readdirSync(storeDir)
+    const oldKey = hashBefore.split('/')[0].replace('.store/', '')
+    t.notOk(storeEntries.includes(oldKey), 'old store entry was cleaned up')
+  }
+})
+
 function setupRequire (cwd) {
   return function requireChain (...chain) {
     return chain.reduce((path, name) => {

